app/src/main/java/clarkson/ee408/tictactoev4/RegisterActivity.java

```diff
@@ -31,40 +42,80 @@ protected void onCreate(Bundle savedInstanceState) {
-        //TODO: set an onclick listener to registerButton to call handleRegister()
-
-        //TODO: set an onclick listener to loginButton to call goBackLogin()
+        registerButton.setOnClickListener(v -> handleRegister());
+        loginButton.setOnClickListener(v -> goBackLogin());
     }
 
     /**
      * Process registration input and pass it to {@link #submitRegistration(User)}
      */
     public void handleRegister() {
-        // TODO: declare local variables for username, password, confirmPassword and displayName. Initialize their values with their corresponding EditText
-
-        // TODO: verify that all fields are not empty before proceeding. Toast with the error message
-
-        // TODO: verify that password is the same af confirm password. Toast with the error message
-
-        // TODO: Create User object with username, display name and password and call submitRegistration()
+        String username = usernameField.getText().toString();
+        String password = passwordField.getText().toString();
+        String confirmPassword = confirmPasswordField.getText().toString();
+        String displayName = displayNameField.getText().toString();
+
+        // verify that all fields are not empty before proceeding
+        if (username.isEmpty() || password.isEmpty() || confirmPassword.isEmpty() || displayName.isEmpty()) {
+            Toast.makeText(this, "fields can not be empty", Toast.LENGTH_SHORT).show();
+            return;
+        }
+
+        // verify that password is the same af confirm password
+        if (!password.equals(confirmPassword)) {
+            Toast.makeText(this, "passwords do not match", Toast.LENGTH_SHORT).show();
+            return;
+        }
+
+        User user = new User(username, displayName, password, false);
+        submitRegistration(user);
     }
 
     /**
      * Sends REGISTER request to the server
      * @param user the User to register
      */
     void submitRegistration(User user) {
-        //TODO: Send a REGISTER request to the server, if SUCCESS reponse, call goBackLogin(). Else, Toast the error message
+        // Send a REGISTER request to the server in a background thread
+        AppExecutors.getInstance().mainThread().execute(()  -> {
+            try {
+                // serialize the user object to JSON
+                String userJson = gson.toJson(user);
+
+                // create a REGISTER request with the user data
+                Request request = new Request(RequestType.REGISTER, userJson);
+
+                // send the request to the server
+                Response response = SocketClient.getInstance().sendRequest(request, Response.class);
+
+                if (response != null && response.getStatus() == ResponseStatus.SUCCESS) {
+                    // if success response, call goBackLogin()
+                    Toast.makeText(RegisterActivity.this, "registration was successful", Toast.LENGTH_SHORT).show();
+                    goBackLogin();
```

Comment:
Now that the outer runnable is being executed in the network thread, we need to make another runnable and do the UI work on the main thread

```suggestion
                    AppExecutors.getInstance().mainThread().execute(()  -> {
                        Toast.makeText(RegisterActivity.this, "Registration was successful", Toast.LENGTH_SHORT).show();
                        goBackLogin();
                    });
```